pkg/devfile/adapters/kubernetes/component/adapter.go

```diff
@@ -240,109 +241,220 @@ func substitueYamlVariables(baseYaml []byte, yamlSubstitutions map[string]string
+					} else if condition["status"] == "True" {
+						success <- watchObject
+					}
+				}
+			}
+		}
+	}()
+
+	select {
+	case val := <-success:
+		return val, nil
+	case err := <-failure:
+		return nil, err
+	case <-time.After(10 * time.Second):
+		return nil, errors.Errorf("timeout while waiting for %s manifest deployment completion", applicationName)
+	}
+}
+
 // Build image for devfile project
 func (a Adapter) Deploy(parameters common.DeployParameters) (err error) {
+
 	namespace := a.Client.Namespace
 	applicationName := a.ComponentName + "-deploy"
 	deploymentManifest := &unstructured.Unstructured{}
 
 	log.Info("\nDeploying manifest")
-	// TODO: Work out how to correctly handle spinners
-	s := log.Spinner("Deploying the manifest")
 
 	// Specify the substitution keys and values
 	yamlSubstitutions := map[string]string{
-		// TODO: this tag is not passed to delete, do we need to template
 		"CONTAINER_IMAGE": parameters.Tag,
-		"PROJECT_NAME":    applicationName,
+		"COMPONENT_NAME":  applicationName,
 		"PORT":            determinePort(parameters.EnvSpecificInfo),
 	}
 
-	// Substitute the values in the manifest file
-	deployYaml := substitueYamlVariables(parameters.ManifestSource, yamlSubstitutions)
-
 	// Build a yaml decoder with the unstructured Scheme
 	yamlDecoder := yaml.NewDecodingSerializer(unstructured.UnstructuredJSONScheme)
 
-	_, gvk, err := yamlDecoder.Decode([]byte(deployYaml), nil, deploymentManifest)
+	// This will override if manifest.yaml is present
+	writtenToManifest := false
+	manifestFile, err := os.Create(filepath.Join(a.Context, ".odo", "manifest.yaml"))
 	if err != nil {
-		return err
+		err = manifestFile.Close()
+		return errors.Wrap(err, "Unable to create the local manifest file")
 	}
 
-	klog.V(3).Infof("Deploy manifest:\n\n%s", deploymentManifest)
-	gvr := schema.GroupVersionResource{Group: gvk.Group, Version: gvk.Version, Resource: strings.ToLower(gvk.Kind + "s")}
-	klog.V(3).Infof("Manifest type: %s", gvr.String())
+	manifests := bytes.Split(parameters.ManifestSource, []byte("---"))
+	for _, manifest := range manifests {
+		if len(manifest) > 0 {
+			// Substitute the values in the manifest file
+			deployYaml := substitueYamlVariables(manifest, yamlSubstitutions)
 
-	labels := map[string]string{
-		"component": applicationName,
-	}
+			_, gvk, err := yamlDecoder.Decode([]byte(deployYaml), nil, deploymentManifest)
+			if err != nil {
+				return errors.Wrap(err, "Failed to decode the manifest yaml")
+			}
 
-	manifestLabels := deploymentManifest.GetLabels()
-	if manifestLabels != nil {
-		for key, value := range labels {
-			manifestLabels[key] = value
-		}
-		deploymentManifest.SetLabels(manifestLabels)
-	} else {
-		deploymentManifest.SetLabels(labels)
-	}
+			deployJSON, err := deploymentManifest.MarshalJSON()
+			if err != nil {
+				return errors.Wrap(err, "Deployment manifest is invalid")
+			}
+			klog.V(3).Infof("Deploy manifest:\n\n%s\n", deployJSON)
```

Comment:
Why do we need it in JSON? It is just to print it out?

Reply:
I was just using that to test that the manifest was valid, and then thought I could print it out.